packages/xsnap/src/xsnap.js

```diff
@@ -344,28 +392,92 @@ export async function xsnap(options) {
    * @returns {AsyncGenerator<Uint8Array, void, undefined>}
    */
   async function* makeSnapshot(description = 'unknown') {
-    // TODO: Refactor to use tmpFile rather than tmpName.
-    const tmpSnapPath = await ptmpName({
-      template: `make-snapshot-${safeHintFromDescription(
-        description,
-      )}-XXXXXX.xss`,
-    });
+    const output = new PassThrough({ highWaterMark: 1024 * 1024 });
+    let piped = false;
+    let cleaned = false;
+    let done = Promise.resolve();
 
+    let snapshotReadSize = 0;
+    let snapshotSize;
     try {
+      /** @type {string} */
+      let snapPath;
+      /** @type {Readable} */
+      let sourceStream;
+
+      const maybePipe = () => {
+        if (!piped && !cleaned) {
+          sourceStream.pipe(output);
+          piped = true;
+        }
+      };
+
+      if (snapshotUseFs) {
+        // TODO: Refactor to use tmpFile rather than tmpName.
+        // eslint-disable-next-line @jessie.js/no-nested-await
+        snapPath = await ptmpName({
+          template: `make-snapshot-${safeHintFromDescription(
+            description,
+          )}-XXXXXX.xss`,
+        });
+        // eslint-disable-next-line @jessie.js/no-nested-await
+        const handle = await fs.open(snapPath, 'w+');
+        // @ts-expect-error 'close' event added in Node 15.4
+        handle.on('close', () => {
+          fs.unlink(snapPath);
+        });
+        sourceStream = handle.createReadStream();
+        finished(output).finally(() => sourceStream.destroy());
+      } else {
+        sourceStream = savedSnapshotsStream;
+        snapPath = '@7';
+        baton.then(maybePipe);
+      }
+
+      const cleanup = () => {
+        if (cleaned) return;
+        cleaned = true;
+        sourceStream.unpipe(output);
+        // eslint-disable-next-line no-use-before-define
+        output.off('data', onData);
+        output.end();
+      };
+      const checkDone = () => {
+        if (snapshotSize !== undefined && !(snapshotReadSize < snapshotSize)) {
```

Comment:
`.. && (snapshotReadSize >= snapshotSize)` would seem more readable, to me

Reply:
Yes but that is not equivalent in the case where `snapshotSize` is `null`.

Reply:
yeah, let's not rely upon `Number(null)===0`: if it's not a number, we shouldn't reach the comparison

Reply:
`isNaN(Number(null)) === true`

Reply:
Used `-1` as that will always be lower than `0` :)